mono/profiler/mono-profiler-malloc.c

```diff
@@ -0,0 +1,1354 @@
+{
+	alloc_lock ();
+
+	if (!outfile) {
+		alloc_unlock ();
+		return;
+	}
+
+	static int dump_count;
+	if (dump_count)
+		fprintf (outfile, ",\n");
+	++dump_count;
+
+	fprintf (outfile, "\t{\n");
+
+	if (log_malloc) {
+		fprintf (outfile, "\t\t\"alloc\": %zu,\n", alloc_bytes);
+		fprintf (outfile, "\t\t\"alloc-count\": %zu,\n", alloc_count);
+		fprintf (outfile, "\t\t\"alloc-waste\": %zu", alloc_waste);
+
+		dump_tag_bag (&malloc_tags, "malloc", 0);
+
+		// Decode function names of allocations that did not get tagged by the runtime and are still alive
+		HT_FOREACH (&alloc_table, AllocInfo, info, {
+			if (!info->tag && !info->alloc_func) {
+				const char *it = NULL;
+				int i;
+				for (i = 0; i < 4; ++i) {
+					Dl_info dl_info;
+					if (!dladdr (info->ips [i], &dl_info))
+						continue;
+					if (strstr (dl_info.dli_sname, "monoeg_")
+						|| strstr (dl_info.dli_sname, "eg_")
+						|| strstr (dl_info.dli_sname, "g_calloc")
+						|| strstr (dl_info.dli_sname, "rehash")
+						|| strstr (dl_info.dli_sname, "g_vasprintf")
+						|| strstr (dl_info.dli_sname, "image_g_malloc0"))
+						continue;
+					it = dl_info.dli_sname;
+					break;
+				}
+				if (it)
+					update_tag (&malloc_tags_by_fun, it, info->size, info->waste);
+				info->alloc_func = it;
+			}
+		});
+
+		dump_tag_bag (&malloc_tags_by_fun, "malloc-by-fun", 0);
+
+		if (STRAY_ALLOCS) {
+			HT_FOREACH (&alloc_table, AllocInfo, info, {
+					if (!info->tag)
+						printf ("stay alloc from %s\n", info->alloc_func);
+			});
+		}
+
+		if (POKE_HASH_TABLES) {
+			HT_FOREACH (&alloc_table, AllocInfo, info, {
+				if (!info->tag || strcmp ("ghashtable", info->tag))
+					continue;
+				printf ("hashtable size %d from %s\n", g_hash_table_size ((GHashTable*)info->node.key), info->alloc_func);
+			});
+		}
+	}
+
+	if (log_memdom) {
+		HT_FOREACH (&memdom_table, MemDomInfo, info, {
+			dump_memdom (info);
+		});
+	}
+
+	if (log_valloc)
+		dump_vmmap ();
+
+	fprintf (outfile, "\n\t}");
+
+
+	alloc_unlock ();
+}
+
+static void
+icall_dump_allocs (void)
+{
+	dump_stats ();
+}
+
+static void
+dump_alloc_stats (void)
+{
+	static int last_time;
+
+	if (!PRINT_ALLOCATION)
+		return;
+	++last_time;
+	dump_stats ();
+}
+
+static void *
+platform_malloc (size_t size)
+{
+	void * res = malloc (size);
+	add_alloc (res, size, NULL);
+
+	return res;
+}
+
+static void *
+platform_realloc (void *mem, size_t count)
+{
+	AllocInfo *old = hashtable_find (&alloc_table, mem);
+	const char *tag = old ? old->tag : NULL;
+
+	del_alloc (mem);
+	void * res = realloc (mem, count);
+	add_alloc (res, count, tag);
+
+	return res;
+ }
+
+static void
+platform_free (void *mem)
+{
+	del_alloc (mem);
+	free (mem);
+}
+
+static void*
+platform_calloc (size_t count, size_t size)
+{
+	void * res = calloc (count, size);
+	add_alloc (res, count * size, NULL);
+
+	return res;
+}
+
+static void
+runtime_inited (MonoProfiler *prof)
+{
+	mono_add_internal_call ("Mono.MemProf::DumpAllocationInfo", icall_dump_allocs);
```

Comment:
There's none. The goal here is to allow users to definer that function and trigger dumps manually.